File: src/com/chbachman/cminus/representation/function/CodeBlockHolder.java
package com.chbachman.cminus.representation.function;

import com.chbachman.cminus.CMinusParser;
import com.chbachman.cminus.representation.Scope;
import com.chbachman.cminus.representation.statement.Statement;

import java.util.List;

/**
 * Created by Chandler on 5/2/17.
 */
public abstract class CodeBlockHolder implements CodeBlock {

    protected List<Statement> statements;

    public CodeBlockHolder() {}

    public CodeBlockHolder(CMinusParser.CodeBlockContext ctx, Scope scope) {
        init(ctx, scope);
    }

    public CodeBlockHolder(List<Statement> statements) {
        this.statements = statements;
    }

    protected void init(CMinusParser.CodeBlockContext ctx, Scope scope) {
        scope.pushScope(this);
        statements = CodeBlock.parse(ctx, scope);
        scope.popScope();
    }

    @Override
    public String middle() {
        StringBuilder b = new StringBuilder();

        for (Statement s : statements) {
            b.append(s.code());
            b.append('\n');
        }

        return b.toString();
    }
}

File: src/main/java/com/chbachman/cminus/representation/function/CodeBlockHolder.kt
package com.chbachman.cminus.representation.function

import com.chbachman.cminus.representation.value.Statement

/**
 * Created by Chandler on 5/2/17.
 * Default Implementation of CodeBlock, handles Middle
 */
abstract class CodeBlockHolder : CodeBlock {

    abstract protected val statements: List<Statement>

    override val middle: String
        get() {
            val b = StringBuilder()

            for (s in statements) {
                b.append(s.statement)
                b.append('\n')
            }

            return b.toString()
        }
}
